{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9f207b61",
   "metadata": {},
   "source": [
    "# PyWGCNA object\n",
    "\n",
    "In this tutorial, we're going to explore the PyWGCNA objects and talk about an important parameters we keep in the objects.\n",
    "\n",
    "you can find all the parameters using [API documentation](https://mortazavilab.github.io/PyWGCNA/html/api.html?highlight=wgcna#PyWGCNA.wgcna.WGCNA).\n",
    "\n",
    "**note**: I would suggest to read the original [WGCNA paper](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559) to become more familiar with the all the main outputs.\n",
    "\n",
    "I'm going to use the PyWGCNA output that we get from Quick_start notebook.\n",
    "\n",
    "### Read PyWGCNA object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "64451149",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[94mReading 5xFAD WGCNA done!\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import PyWGCNA\n",
    "pyWGCNA_5xFAD = PyWGCNA.readWGCNA(\"5xFAD.p\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30b54953",
   "metadata": {},
   "source": [
    "### Raw expression data\n",
    "\n",
    "The raw expression data store in `geneExpr` in AnnData format which rows/obs are samples/sample information and cols/var are genes/gene information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "637dedc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw expresion data along with information:\n",
      " AnnData object with n_obs × n_vars = 192 × 55448\n",
      "    obs: 'Age', 'Tissue', 'Sex', 'Genotype'\n",
      "    var: 'gene_name', 'gene_biotype'\n"
     ]
    }
   ],
   "source": [
    "print(f\"Raw expresion data along with information:\\n {pyWGCNA_5xFAD.geneExpr}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6617df78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw sample inforamtions:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Age</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Genotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X4mo_cortex_F_5xFADHEMI_430</th>\n",
       "      <td>4mon</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>Female</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X4mo_cortex_F_5xFADHEMI_431</th>\n",
       "      <td>4mon</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>Female</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X4mo_cortex_F_5xFADHEMI_433</th>\n",
       "      <td>4mon</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>Female</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X4mo_cortex_F_5xFADHEMI_434</th>\n",
       "      <td>4mon</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>Female</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X4mo_cortex_F_5xFADHEMI_511</th>\n",
       "      <td>4mon</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>Female</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Age  Tissue     Sex   Genotype\n",
       "sample_id                                                   \n",
       "X4mo_cortex_F_5xFADHEMI_430  4mon  Cortex  Female  5xFADHEMI\n",
       "X4mo_cortex_F_5xFADHEMI_431  4mon  Cortex  Female  5xFADHEMI\n",
       "X4mo_cortex_F_5xFADHEMI_433  4mon  Cortex  Female  5xFADHEMI\n",
       "X4mo_cortex_F_5xFADHEMI_434  4mon  Cortex  Female  5xFADHEMI\n",
       "X4mo_cortex_F_5xFADHEMI_511  4mon  Cortex  Female  5xFADHEMI"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Raw sample inforamtions:\")\n",
    "pyWGCNA_5xFAD.geneExpr.obs.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4be2c046",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw gene inforamtions:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>Pbsn</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>Cdc45</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>H19</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>Scml2</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049</th>\n",
       "      <td>Apoh</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_name    gene_biotype\n",
       "ENSMUSG00000000003      Pbsn  protein_coding\n",
       "ENSMUSG00000000028     Cdc45  protein_coding\n",
       "ENSMUSG00000000031       H19          lncRNA\n",
       "ENSMUSG00000000037     Scml2  protein_coding\n",
       "ENSMUSG00000000049      Apoh  protein_coding"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Raw gene inforamtions:\")\n",
    "pyWGCNA_5xFAD.geneExpr.var.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c23daba5",
   "metadata": {},
   "source": [
    "### Processed expresion matrix\n",
    "\n",
    "After you ran PyWGCNA, you can use `datExpr` to look into results. The structure is similar to `geneExpr` which is AnnData format which rows/obs are samples/sample information and cols/var are genes/gene information. The main differences is now we have all the modules information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "484c336a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed expresion data along with information:\n",
      " AnnData object with n_obs × n_vars = 192 × 23844\n",
      "    obs: 'Age', 'Tissue', 'Sex', 'Genotype'\n",
      "    var: 'dynamicColors', 'moduleColors', 'moduleLabels', 'gene_name', 'gene_biotype'\n"
     ]
    }
   ],
   "source": [
    "print(f\"Processed expresion data along with information:\\n {pyWGCNA_5xFAD.datExpr}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "15301262",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw sample inforamtions:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Age</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Genotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X4mo_cortex_F_5xFADHEMI_430</th>\n",
       "      <td>4mon</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>Female</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X4mo_cortex_F_5xFADHEMI_431</th>\n",
       "      <td>4mon</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>Female</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X4mo_cortex_F_5xFADHEMI_433</th>\n",
       "      <td>4mon</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>Female</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X4mo_cortex_F_5xFADHEMI_434</th>\n",
       "      <td>4mon</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>Female</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X4mo_cortex_F_5xFADHEMI_511</th>\n",
       "      <td>4mon</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>Female</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Age  Tissue     Sex   Genotype\n",
       "sample_id                                                   \n",
       "X4mo_cortex_F_5xFADHEMI_430  4mon  Cortex  Female  5xFADHEMI\n",
       "X4mo_cortex_F_5xFADHEMI_431  4mon  Cortex  Female  5xFADHEMI\n",
       "X4mo_cortex_F_5xFADHEMI_433  4mon  Cortex  Female  5xFADHEMI\n",
       "X4mo_cortex_F_5xFADHEMI_434  4mon  Cortex  Female  5xFADHEMI\n",
       "X4mo_cortex_F_5xFADHEMI_511  4mon  Cortex  Female  5xFADHEMI"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Raw sample inforamtions:\")\n",
    "pyWGCNA_5xFAD.datExpr.obs.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1cc6b494",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw gene inforamtions:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dynamicColors</th>\n",
       "      <th>moduleColors</th>\n",
       "      <th>moduleLabels</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>darkred</td>\n",
       "      <td>darkred</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049</th>\n",
       "      <td>darkred</td>\n",
       "      <td>darkred</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Apoh</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000056</th>\n",
       "      <td>darkgrey</td>\n",
       "      <td>darkgrey</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Narf</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000058</th>\n",
       "      <td>coral</td>\n",
       "      <td>coral</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Cav2</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000078</th>\n",
       "      <td>gainsboro</td>\n",
       "      <td>gainsboro</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Klf6</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   dynamicColors moduleColors  moduleLabels gene_name  \\\n",
       "ENSMUSG00000000028       darkred      darkred           4.0     Cdc45   \n",
       "ENSMUSG00000000049       darkred      darkred           4.0      Apoh   \n",
       "ENSMUSG00000000056      darkgrey     darkgrey           3.0      Narf   \n",
       "ENSMUSG00000000058         coral        coral           2.0      Cav2   \n",
       "ENSMUSG00000000078     gainsboro    gainsboro           7.0      Klf6   \n",
       "\n",
       "                      gene_biotype  \n",
       "ENSMUSG00000000028  protein_coding  \n",
       "ENSMUSG00000000049  protein_coding  \n",
       "ENSMUSG00000000056  protein_coding  \n",
       "ENSMUSG00000000058  protein_coding  \n",
       "ENSMUSG00000000078  protein_coding  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Raw gene inforamtions:\")\n",
    "pyWGCNA_5xFAD.datExpr.var.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78d8e4fa",
   "metadata": {},
   "source": [
    "### Updating gene information\n",
    "\n",
    "if you want to update gene inforamtion, you can use [`updateGeneInfo()` function](https://mortazavilab.github.io/PyWGCNA/html/api.html?highlight=updategeneinfo#PyWGCNA.wgcna.WGCNA.updateGeneInfo).\n",
    "\n",
    "**note**: For doing functional enrichment analysis, you should have your gene name as a column (named should be gene_name) in your gene information dataframe.\n",
    "\n",
    "[`getGeneList()` function](https://mortazavilab.github.io/PyWGCNA/html/api.html?highlight=getgenelist#PyWGCNA.utils.getGeneList) is helpping you to get gene name and all the other information you want to have like gene biotype using [biomart from ensemble](https://www.ensembl.org/biomart/martview/925c00ece2a04471e061de6e720c5cc7)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3568bfbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "geneList = PyWGCNA.getGeneList(dataset='mmusculus_gene_ensembl',\n",
    "                               attributes=['ensembl_gene_id', \n",
    "                                           'external_gene_name', \n",
    "                                           'gene_biotype'],\n",
    "                               maps=['gene_id', 'gene_name', 'gene_biotype'])\n",
    "\n",
    "\n",
    "pyWGCNA_5xFAD.updateGeneInfo(geneList)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b0fc1d9",
   "metadata": {},
   "source": [
    "### Updating sample information\n",
    "\n",
    "You can also update your sample information by using [`updateSampleInfo()` function](https://mortazavilab.github.io/PyWGCNA/html/api.html?highlight=updatesampleinfo#PyWGCNA.wgcna.WGCNA.updateSampleInfo)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00951147",
   "metadata": {},
   "outputs": [],
   "source": [
    "pyWGCNA_5xFAD.updateSampleInfo(path='5xFAD_paper/sampleInfo.csv', sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ceba9dfa",
   "metadata": {},
   "source": [
    "### Soft power\n",
    "\n",
    "if you want to see the values for each power it stores in the `sft`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f35292ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Power</th>\n",
       "      <th>SFT.R.sq</th>\n",
       "      <th>slope</th>\n",
       "      <th>truncated R.sq</th>\n",
       "      <th>mean(k)</th>\n",
       "      <th>median(k)</th>\n",
       "      <th>max(k)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.368857</td>\n",
       "      <td>-0.481613</td>\n",
       "      <td>0.701585</td>\n",
       "      <td>2444.750755</td>\n",
       "      <td>2260.416617</td>\n",
       "      <td>5665.102677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>0.7253</td>\n",
       "      <td>-0.99165</td>\n",
       "      <td>0.886361</td>\n",
       "      <td>840.665489</td>\n",
       "      <td>673.081242</td>\n",
       "      <td>3009.058817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>0.791986</td>\n",
       "      <td>-1.194264</td>\n",
       "      <td>0.946969</td>\n",
       "      <td>385.685335</td>\n",
       "      <td>258.451266</td>\n",
       "      <td>1916.810604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>0.835392</td>\n",
       "      <td>-1.3419</td>\n",
       "      <td>0.968446</td>\n",
       "      <td>207.404152</td>\n",
       "      <td>113.456084</td>\n",
       "      <td>1332.762767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>0.853842</td>\n",
       "      <td>-1.472183</td>\n",
       "      <td>0.973346</td>\n",
       "      <td>123.232581</td>\n",
       "      <td>54.784481</td>\n",
       "      <td>984.03682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>0.870673</td>\n",
       "      <td>-1.553348</td>\n",
       "      <td>0.979584</td>\n",
       "      <td>78.455923</td>\n",
       "      <td>28.471239</td>\n",
       "      <td>752.959995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>0.886736</td>\n",
       "      <td>-1.600869</td>\n",
       "      <td>0.986635</td>\n",
       "      <td>52.572016</td>\n",
       "      <td>15.594822</td>\n",
       "      <td>591.514189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>0.896672</td>\n",
       "      <td>-1.639343</td>\n",
       "      <td>0.992373</td>\n",
       "      <td>36.65884</td>\n",
       "      <td>9.454046</td>\n",
       "      <td>475.817179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>0.903531</td>\n",
       "      <td>-1.677747</td>\n",
       "      <td>0.994643</td>\n",
       "      <td>26.397061</td>\n",
       "      <td>6.02443</td>\n",
       "      <td>389.237529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>0.906045</td>\n",
       "      <td>-1.706474</td>\n",
       "      <td>0.995895</td>\n",
       "      <td>19.521431</td>\n",
       "      <td>3.975959</td>\n",
       "      <td>322.823835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>0.905582</td>\n",
       "      <td>-1.731076</td>\n",
       "      <td>0.994806</td>\n",
       "      <td>14.767291</td>\n",
       "      <td>2.623921</td>\n",
       "      <td>270.867414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>13</td>\n",
       "      <td>0.914482</td>\n",
       "      <td>-1.751347</td>\n",
       "      <td>0.997466</td>\n",
       "      <td>8.941254</td>\n",
       "      <td>1.205108</td>\n",
       "      <td>196.222412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>15</td>\n",
       "      <td>0.912684</td>\n",
       "      <td>-1.771227</td>\n",
       "      <td>0.994189</td>\n",
       "      <td>5.759987</td>\n",
       "      <td>0.568044</td>\n",
       "      <td>146.575347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>17</td>\n",
       "      <td>0.912188</td>\n",
       "      <td>-1.774908</td>\n",
       "      <td>0.990829</td>\n",
       "      <td>3.905403</td>\n",
       "      <td>0.273242</td>\n",
       "      <td>112.189051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>19</td>\n",
       "      <td>0.907649</td>\n",
       "      <td>-1.774186</td>\n",
       "      <td>0.989457</td>\n",
       "      <td>2.766824</td>\n",
       "      <td>0.135454</td>\n",
       "      <td>87.594343</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Power  SFT.R.sq     slope truncated R.sq      mean(k)    median(k)  \\\n",
       "0       1  0.368857 -0.481613       0.701585  2444.750755  2260.416617   \n",
       "1       2    0.7253  -0.99165       0.886361   840.665489   673.081242   \n",
       "2       3  0.791986 -1.194264       0.946969   385.685335   258.451266   \n",
       "3       4  0.835392   -1.3419       0.968446   207.404152   113.456084   \n",
       "4       5  0.853842 -1.472183       0.973346   123.232581    54.784481   \n",
       "5       6  0.870673 -1.553348       0.979584    78.455923    28.471239   \n",
       "6       7  0.886736 -1.600869       0.986635    52.572016    15.594822   \n",
       "7       8  0.896672 -1.639343       0.992373     36.65884     9.454046   \n",
       "8       9  0.903531 -1.677747       0.994643    26.397061      6.02443   \n",
       "9      10  0.906045 -1.706474       0.995895    19.521431     3.975959   \n",
       "10     11  0.905582 -1.731076       0.994806    14.767291     2.623921   \n",
       "11     13  0.914482 -1.751347       0.997466     8.941254     1.205108   \n",
       "12     15  0.912684 -1.771227       0.994189     5.759987     0.568044   \n",
       "13     17  0.912188 -1.774908       0.990829     3.905403     0.273242   \n",
       "14     19  0.907649 -1.774186       0.989457     2.766824     0.135454   \n",
       "\n",
       "         max(k)  \n",
       "0   5665.102677  \n",
       "1   3009.058817  \n",
       "2   1916.810604  \n",
       "3   1332.762767  \n",
       "4     984.03682  \n",
       "5    752.959995  \n",
       "6    591.514189  \n",
       "7    475.817179  \n",
       "8    389.237529  \n",
       "9    322.823835  \n",
       "10   270.867414  \n",
       "11   196.222412  \n",
       "12   146.575347  \n",
       "13   112.189051  \n",
       "14    87.594343  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyWGCNA_5xFAD.sft"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80e0dda8",
   "metadata": {},
   "source": [
    "### adjacency matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2e8c813b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <th>ENSMUSG00000000049</th>\n",
       "      <th>ENSMUSG00000000056</th>\n",
       "      <th>ENSMUSG00000000058</th>\n",
       "      <th>ENSMUSG00000000078</th>\n",
       "      <th>ENSMUSG00000000085</th>\n",
       "      <th>ENSMUSG00000000088</th>\n",
       "      <th>ENSMUSG00000000120</th>\n",
       "      <th>ENSMUSG00000000125</th>\n",
       "      <th>ENSMUSG00000000126</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSMUSG00000118346</th>\n",
       "      <th>ENSMUSG00000118353</th>\n",
       "      <th>ENSMUSG00000118369</th>\n",
       "      <th>ENSMUSG00000118370</th>\n",
       "      <th>ENSMUSG00000118371</th>\n",
       "      <th>ENSMUSG00000118372</th>\n",
       "      <th>ENSMUSG00000118380</th>\n",
       "      <th>ENSMUSG00000118382</th>\n",
       "      <th>ENSMUSG00000118383</th>\n",
       "      <th>ENSMUSG00000118385</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.679174e-08</td>\n",
       "      <td>3.268029e-05</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.420587e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.328571e-10</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>5.694075e-11</td>\n",
       "      <td>2.787736e-09</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.583412e-17</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.632674e-10</td>\n",
       "      <td>1.313950e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049</th>\n",
       "      <td>2.679174e-08</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.913747e-21</td>\n",
       "      <td>9.101099e-11</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.715985e-11</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.534340e-08</td>\n",
       "      <td>2.247703e-16</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.869302e-18</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.019448e-18</td>\n",
       "      <td>1.444743e-12</td>\n",
       "      <td>3.035505e-06</td>\n",
       "      <td>3.452410e-09</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000056</th>\n",
       "      <td>3.268029e-05</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.179837e-02</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.572518e-08</td>\n",
       "      <td>4.264834e-09</td>\n",
       "      <td>6.374569e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.321102e-08</td>\n",
       "      <td>8.969329e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.178820e-21</td>\n",
       "      <td>6.750673e-13</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.789749e-06</td>\n",
       "      <td>3.403363e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000058</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.913747e-21</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.049737e-05</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.152316e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>3.042050e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.111779e-16</td>\n",
       "      <td>3.390977e-12</td>\n",
       "      <td>5.422329e-12</td>\n",
       "      <td>2.351625e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.179132e-09</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000078</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>9.101099e-11</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.049737e-05</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.807784e-03</td>\n",
       "      <td>8.560779e-12</td>\n",
       "      <td>2.072939e-13</td>\n",
       "      <td>3.573166e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.323349e-20</td>\n",
       "      <td>1.247663e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>8.545731e-11</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118372</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.019448e-18</td>\n",
       "      <td>1.178820e-21</td>\n",
       "      <td>2.351625e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.491032e-28</td>\n",
       "      <td>1.109072e-16</td>\n",
       "      <td>4.666838e-20</td>\n",
       "      <td>5.265967e-21</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.785356e-16</td>\n",
       "      <td>3.314043e-19</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.312963e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.108374e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118380</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.444743e-12</td>\n",
       "      <td>6.750673e-13</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.270510e-19</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.932149e-13</td>\n",
       "      <td>6.237386e-11</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>4.069893e-09</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.241742e-25</td>\n",
       "      <td>1.331749e-13</td>\n",
       "      <td>3.312963e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>7.789279e-12</td>\n",
       "      <td>1.956529e-15</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118382</th>\n",
       "      <td>5.632674e-10</td>\n",
       "      <td>3.035505e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.179132e-09</td>\n",
       "      <td>8.545731e-11</td>\n",
       "      <td>1.691904e-09</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.742938e-13</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>5.938378e-11</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.182987e-13</td>\n",
       "      <td>9.909548e-10</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>7.789279e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.881980e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118383</th>\n",
       "      <td>1.313950e-06</td>\n",
       "      <td>3.452410e-09</td>\n",
       "      <td>2.789749e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.894810e-04</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.957643e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.195629e-10</td>\n",
       "      <td>3.178509e-08</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.874160e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.956529e-15</td>\n",
       "      <td>5.881980e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118385</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.403363e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.761458e-09</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.103193e-10</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.938846e-10</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.203996e-14</td>\n",
       "      <td>1.484882e-09</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.108374e-12</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23844 rows × 23844 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ENSMUSG00000000028  ENSMUSG00000000049  \\\n",
       "ENSMUSG00000000028        0.000000e+00        2.679174e-08   \n",
       "ENSMUSG00000000049        2.679174e-08        0.000000e+00   \n",
       "ENSMUSG00000000056        3.268029e-05        0.000000e+00   \n",
       "ENSMUSG00000000058        0.000000e+00        6.913747e-21   \n",
       "ENSMUSG00000000078        0.000000e+00        9.101099e-11   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        0.000000e+00        3.019448e-18   \n",
       "ENSMUSG00000118380        0.000000e+00        1.444743e-12   \n",
       "ENSMUSG00000118382        5.632674e-10        3.035505e-06   \n",
       "ENSMUSG00000118383        1.313950e-06        3.452410e-09   \n",
       "ENSMUSG00000118385        0.000000e+00        0.000000e+00   \n",
       "\n",
       "                    ENSMUSG00000000056  ENSMUSG00000000058  \\\n",
       "ENSMUSG00000000028        3.268029e-05        0.000000e+00   \n",
       "ENSMUSG00000000049        0.000000e+00        6.913747e-21   \n",
       "ENSMUSG00000000056        0.000000e+00        0.000000e+00   \n",
       "ENSMUSG00000000058        0.000000e+00        0.000000e+00   \n",
       "ENSMUSG00000000078        0.000000e+00        4.049737e-05   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        1.178820e-21        2.351625e-12   \n",
       "ENSMUSG00000118380        6.750673e-13        0.000000e+00   \n",
       "ENSMUSG00000118382        0.000000e+00        1.179132e-09   \n",
       "ENSMUSG00000118383        2.789749e-06        0.000000e+00   \n",
       "ENSMUSG00000118385        3.403363e-07        0.000000e+00   \n",
       "\n",
       "                    ENSMUSG00000000078  ENSMUSG00000000085  \\\n",
       "ENSMUSG00000000028        0.000000e+00        1.420587e-06   \n",
       "ENSMUSG00000000049        9.101099e-11        0.000000e+00   \n",
       "ENSMUSG00000000056        0.000000e+00        1.179837e-02   \n",
       "ENSMUSG00000000058        4.049737e-05        0.000000e+00   \n",
       "ENSMUSG00000000078        0.000000e+00        0.000000e+00   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        0.000000e+00        0.000000e+00   \n",
       "ENSMUSG00000118380        0.000000e+00        2.270510e-19   \n",
       "ENSMUSG00000118382        8.545731e-11        1.691904e-09   \n",
       "ENSMUSG00000118383        0.000000e+00        6.894810e-04   \n",
       "ENSMUSG00000118385        0.000000e+00        1.761458e-09   \n",
       "\n",
       "                    ENSMUSG00000000088  ENSMUSG00000000120  \\\n",
       "ENSMUSG00000000028        0.000000e+00        0.000000e+00   \n",
       "ENSMUSG00000000049        4.715985e-11        0.000000e+00   \n",
       "ENSMUSG00000000056        0.000000e+00        5.572518e-08   \n",
       "ENSMUSG00000000058        3.152316e-07        0.000000e+00   \n",
       "ENSMUSG00000000078        4.807784e-03        8.560779e-12   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        5.491032e-28        1.109072e-16   \n",
       "ENSMUSG00000118380        0.000000e+00        6.932149e-13   \n",
       "ENSMUSG00000118382        0.000000e+00        0.000000e+00   \n",
       "ENSMUSG00000118383        0.000000e+00        0.000000e+00   \n",
       "ENSMUSG00000118385        0.000000e+00        3.103193e-10   \n",
       "\n",
       "                    ENSMUSG00000000125  ENSMUSG00000000126  ...  \\\n",
       "ENSMUSG00000000028        1.328571e-10        0.000000e+00  ...   \n",
       "ENSMUSG00000000049        6.534340e-08        2.247703e-16  ...   \n",
       "ENSMUSG00000000056        4.264834e-09        6.374569e-07  ...   \n",
       "ENSMUSG00000000058        0.000000e+00        0.000000e+00  ...   \n",
       "ENSMUSG00000000078        2.072939e-13        3.573166e-05  ...   \n",
       "...                                ...                 ...  ...   \n",
       "ENSMUSG00000118372        4.666838e-20        5.265967e-21  ...   \n",
       "ENSMUSG00000118380        6.237386e-11        0.000000e+00  ...   \n",
       "ENSMUSG00000118382        1.742938e-13        0.000000e+00  ...   \n",
       "ENSMUSG00000118383        5.957643e-07        0.000000e+00  ...   \n",
       "ENSMUSG00000118385        0.000000e+00        2.938846e-10  ...   \n",
       "\n",
       "                    ENSMUSG00000118346  ENSMUSG00000118353  \\\n",
       "ENSMUSG00000000028        5.694075e-11        2.787736e-09   \n",
       "ENSMUSG00000000049        0.000000e+00        0.000000e+00   \n",
       "ENSMUSG00000000056        0.000000e+00        3.321102e-08   \n",
       "ENSMUSG00000000058        3.042050e-07        0.000000e+00   \n",
       "ENSMUSG00000000078        0.000000e+00        0.000000e+00   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        0.000000e+00        5.785356e-16   \n",
       "ENSMUSG00000118380        4.069893e-09        0.000000e+00   \n",
       "ENSMUSG00000118382        5.938378e-11        0.000000e+00   \n",
       "ENSMUSG00000118383        1.195629e-10        3.178509e-08   \n",
       "ENSMUSG00000118385        0.000000e+00        1.203996e-14   \n",
       "\n",
       "                    ENSMUSG00000118369  ENSMUSG00000118370  \\\n",
       "ENSMUSG00000000028        0.000000e+00        0.000000e+00   \n",
       "ENSMUSG00000000049        3.869302e-18        0.000000e+00   \n",
       "ENSMUSG00000000056        8.969329e-07        0.000000e+00   \n",
       "ENSMUSG00000000058        1.111779e-16        3.390977e-12   \n",
       "ENSMUSG00000000078        2.323349e-20        1.247663e-07   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        3.314043e-19        0.000000e+00   \n",
       "ENSMUSG00000118380        0.000000e+00        6.241742e-25   \n",
       "ENSMUSG00000118382        0.000000e+00        2.182987e-13   \n",
       "ENSMUSG00000118383        0.000000e+00        0.000000e+00   \n",
       "ENSMUSG00000118385        1.484882e-09        0.000000e+00   \n",
       "\n",
       "                    ENSMUSG00000118371  ENSMUSG00000118372  \\\n",
       "ENSMUSG00000000028        2.583412e-17        0.000000e+00   \n",
       "ENSMUSG00000000049        0.000000e+00        3.019448e-18   \n",
       "ENSMUSG00000000056        0.000000e+00        1.178820e-21   \n",
       "ENSMUSG00000000058        5.422329e-12        2.351625e-12   \n",
       "ENSMUSG00000000078        0.000000e+00        0.000000e+00   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        0.000000e+00        0.000000e+00   \n",
       "ENSMUSG00000118380        1.331749e-13        3.312963e-12   \n",
       "ENSMUSG00000118382        9.909548e-10        0.000000e+00   \n",
       "ENSMUSG00000118383        1.874160e-12        0.000000e+00   \n",
       "ENSMUSG00000118385        0.000000e+00        4.108374e-12   \n",
       "\n",
       "                    ENSMUSG00000118380  ENSMUSG00000118382  \\\n",
       "ENSMUSG00000000028        0.000000e+00        5.632674e-10   \n",
       "ENSMUSG00000000049        1.444743e-12        3.035505e-06   \n",
       "ENSMUSG00000000056        6.750673e-13        0.000000e+00   \n",
       "ENSMUSG00000000058        0.000000e+00        1.179132e-09   \n",
       "ENSMUSG00000000078        0.000000e+00        8.545731e-11   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        3.312963e-12        0.000000e+00   \n",
       "ENSMUSG00000118380        0.000000e+00        7.789279e-12   \n",
       "ENSMUSG00000118382        7.789279e-12        0.000000e+00   \n",
       "ENSMUSG00000118383        1.956529e-15        5.881980e-07   \n",
       "ENSMUSG00000118385        0.000000e+00        0.000000e+00   \n",
       "\n",
       "                    ENSMUSG00000118383  ENSMUSG00000118385  \n",
       "ENSMUSG00000000028        1.313950e-06        0.000000e+00  \n",
       "ENSMUSG00000000049        3.452410e-09        0.000000e+00  \n",
       "ENSMUSG00000000056        2.789749e-06        3.403363e-07  \n",
       "ENSMUSG00000000058        0.000000e+00        0.000000e+00  \n",
       "ENSMUSG00000000078        0.000000e+00        0.000000e+00  \n",
       "...                                ...                 ...  \n",
       "ENSMUSG00000118372        0.000000e+00        4.108374e-12  \n",
       "ENSMUSG00000118380        1.956529e-15        0.000000e+00  \n",
       "ENSMUSG00000118382        5.881980e-07        0.000000e+00  \n",
       "ENSMUSG00000118383        0.000000e+00        0.000000e+00  \n",
       "ENSMUSG00000118385        0.000000e+00        0.000000e+00  \n",
       "\n",
       "[23844 rows x 23844 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyWGCNA_5xFAD.adjacency"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af22fe32",
   "metadata": {},
   "source": [
    "### Topological Overlap Matrix (TOM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "555a32de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <th>ENSMUSG00000000049</th>\n",
       "      <th>ENSMUSG00000000056</th>\n",
       "      <th>ENSMUSG00000000058</th>\n",
       "      <th>ENSMUSG00000000078</th>\n",
       "      <th>ENSMUSG00000000085</th>\n",
       "      <th>ENSMUSG00000000088</th>\n",
       "      <th>ENSMUSG00000000120</th>\n",
       "      <th>ENSMUSG00000000125</th>\n",
       "      <th>ENSMUSG00000000126</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSMUSG00000118346</th>\n",
       "      <th>ENSMUSG00000118353</th>\n",
       "      <th>ENSMUSG00000118369</th>\n",
       "      <th>ENSMUSG00000118370</th>\n",
       "      <th>ENSMUSG00000118371</th>\n",
       "      <th>ENSMUSG00000118372</th>\n",
       "      <th>ENSMUSG00000118380</th>\n",
       "      <th>ENSMUSG00000118382</th>\n",
       "      <th>ENSMUSG00000118383</th>\n",
       "      <th>ENSMUSG00000118385</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.737436e-08</td>\n",
       "      <td>1.547511e-03</td>\n",
       "      <td>2.073630e-08</td>\n",
       "      <td>4.725714e-09</td>\n",
       "      <td>2.397450e-03</td>\n",
       "      <td>3.166083e-09</td>\n",
       "      <td>1.218259e-07</td>\n",
       "      <td>5.862595e-07</td>\n",
       "      <td>5.512740e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>1.396855e-05</td>\n",
       "      <td>3.388088e-08</td>\n",
       "      <td>4.887126e-07</td>\n",
       "      <td>5.985277e-12</td>\n",
       "      <td>1.366288e-09</td>\n",
       "      <td>3.715675e-10</td>\n",
       "      <td>8.465453e-10</td>\n",
       "      <td>5.111236e-07</td>\n",
       "      <td>2.642267e-05</td>\n",
       "      <td>7.450698e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049</th>\n",
       "      <td>2.737436e-08</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>6.104462e-08</td>\n",
       "      <td>3.473789e-09</td>\n",
       "      <td>8.791817e-08</td>\n",
       "      <td>7.308956e-08</td>\n",
       "      <td>5.519291e-07</td>\n",
       "      <td>5.496296e-10</td>\n",
       "      <td>1.428166e-07</td>\n",
       "      <td>4.770658e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>1.827312e-08</td>\n",
       "      <td>1.848953e-12</td>\n",
       "      <td>9.663349e-09</td>\n",
       "      <td>2.443669e-10</td>\n",
       "      <td>3.150939e-10</td>\n",
       "      <td>1.712946e-10</td>\n",
       "      <td>3.459822e-10</td>\n",
       "      <td>3.414124e-06</td>\n",
       "      <td>6.195964e-09</td>\n",
       "      <td>1.261908e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000056</th>\n",
       "      <td>1.547511e-03</td>\n",
       "      <td>6.104462e-08</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>9.395216e-07</td>\n",
       "      <td>1.383738e-07</td>\n",
       "      <td>5.168019e-02</td>\n",
       "      <td>6.775585e-10</td>\n",
       "      <td>1.965049e-04</td>\n",
       "      <td>1.342360e-04</td>\n",
       "      <td>8.450327e-04</td>\n",
       "      <td>...</td>\n",
       "      <td>6.585115e-06</td>\n",
       "      <td>2.099789e-05</td>\n",
       "      <td>8.782336e-04</td>\n",
       "      <td>1.147629e-09</td>\n",
       "      <td>8.615273e-08</td>\n",
       "      <td>5.748162e-09</td>\n",
       "      <td>8.779400e-08</td>\n",
       "      <td>7.341468e-06</td>\n",
       "      <td>4.213346e-03</td>\n",
       "      <td>2.630794e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000058</th>\n",
       "      <td>2.073630e-08</td>\n",
       "      <td>3.473789e-09</td>\n",
       "      <td>9.395216e-07</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.402169e-04</td>\n",
       "      <td>4.533858e-07</td>\n",
       "      <td>3.763688e-04</td>\n",
       "      <td>9.523902e-08</td>\n",
       "      <td>1.230038e-08</td>\n",
       "      <td>1.043700e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>5.468722e-03</td>\n",
       "      <td>1.308723e-09</td>\n",
       "      <td>9.309275e-07</td>\n",
       "      <td>1.077963e-08</td>\n",
       "      <td>5.672777e-09</td>\n",
       "      <td>1.063537e-09</td>\n",
       "      <td>3.631234e-10</td>\n",
       "      <td>1.143298e-05</td>\n",
       "      <td>1.094910e-07</td>\n",
       "      <td>1.946673e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000078</th>\n",
       "      <td>4.725714e-09</td>\n",
       "      <td>8.791817e-08</td>\n",
       "      <td>1.383738e-07</td>\n",
       "      <td>1.402169e-04</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.399752e-05</td>\n",
       "      <td>6.002748e-03</td>\n",
       "      <td>4.886556e-05</td>\n",
       "      <td>1.993526e-06</td>\n",
       "      <td>3.175543e-03</td>\n",
       "      <td>...</td>\n",
       "      <td>1.299579e-07</td>\n",
       "      <td>8.334004e-13</td>\n",
       "      <td>5.657249e-05</td>\n",
       "      <td>3.574995e-07</td>\n",
       "      <td>9.083470e-09</td>\n",
       "      <td>9.185583e-10</td>\n",
       "      <td>1.696247e-09</td>\n",
       "      <td>2.139086e-05</td>\n",
       "      <td>4.533132e-06</td>\n",
       "      <td>2.557257e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118372</th>\n",
       "      <td>3.715675e-10</td>\n",
       "      <td>1.712946e-10</td>\n",
       "      <td>5.748162e-09</td>\n",
       "      <td>1.063537e-09</td>\n",
       "      <td>9.185583e-10</td>\n",
       "      <td>3.497042e-09</td>\n",
       "      <td>2.280045e-09</td>\n",
       "      <td>1.969702e-09</td>\n",
       "      <td>3.565670e-11</td>\n",
       "      <td>3.751951e-09</td>\n",
       "      <td>...</td>\n",
       "      <td>5.720627e-10</td>\n",
       "      <td>8.491071e-09</td>\n",
       "      <td>2.053632e-09</td>\n",
       "      <td>1.828569e-10</td>\n",
       "      <td>4.668191e-12</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.677540e-10</td>\n",
       "      <td>2.121388e-11</td>\n",
       "      <td>9.607878e-11</td>\n",
       "      <td>9.978972e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118380</th>\n",
       "      <td>8.465453e-10</td>\n",
       "      <td>3.459822e-10</td>\n",
       "      <td>8.779400e-08</td>\n",
       "      <td>3.631234e-10</td>\n",
       "      <td>1.696247e-09</td>\n",
       "      <td>2.325875e-07</td>\n",
       "      <td>6.027817e-09</td>\n",
       "      <td>2.778879e-08</td>\n",
       "      <td>1.148290e-08</td>\n",
       "      <td>4.950899e-12</td>\n",
       "      <td>...</td>\n",
       "      <td>5.866245e-07</td>\n",
       "      <td>1.493508e-10</td>\n",
       "      <td>7.485294e-11</td>\n",
       "      <td>5.500307e-10</td>\n",
       "      <td>9.493952e-11</td>\n",
       "      <td>1.677540e-10</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.195543e-06</td>\n",
       "      <td>8.592797e-08</td>\n",
       "      <td>3.328797e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118382</th>\n",
       "      <td>5.111236e-07</td>\n",
       "      <td>3.414124e-06</td>\n",
       "      <td>7.341468e-06</td>\n",
       "      <td>1.143298e-05</td>\n",
       "      <td>2.139086e-05</td>\n",
       "      <td>7.657257e-05</td>\n",
       "      <td>1.623556e-05</td>\n",
       "      <td>2.521969e-09</td>\n",
       "      <td>6.758380e-08</td>\n",
       "      <td>5.919615e-09</td>\n",
       "      <td>...</td>\n",
       "      <td>9.139048e-06</td>\n",
       "      <td>1.424658e-08</td>\n",
       "      <td>1.356658e-08</td>\n",
       "      <td>2.996882e-08</td>\n",
       "      <td>2.540177e-07</td>\n",
       "      <td>2.121388e-11</td>\n",
       "      <td>3.195543e-06</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>6.855203e-05</td>\n",
       "      <td>1.614268e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118383</th>\n",
       "      <td>2.642267e-05</td>\n",
       "      <td>6.195964e-09</td>\n",
       "      <td>4.213346e-03</td>\n",
       "      <td>1.094910e-07</td>\n",
       "      <td>4.533132e-06</td>\n",
       "      <td>1.784947e-02</td>\n",
       "      <td>1.140550e-09</td>\n",
       "      <td>2.053265e-07</td>\n",
       "      <td>1.130612e-05</td>\n",
       "      <td>1.774412e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>1.389545e-05</td>\n",
       "      <td>6.388611e-07</td>\n",
       "      <td>7.089825e-07</td>\n",
       "      <td>2.036567e-10</td>\n",
       "      <td>9.645142e-08</td>\n",
       "      <td>9.607878e-11</td>\n",
       "      <td>8.592797e-08</td>\n",
       "      <td>6.855203e-05</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.113636e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118385</th>\n",
       "      <td>7.450698e-09</td>\n",
       "      <td>1.261908e-13</td>\n",
       "      <td>2.630794e-05</td>\n",
       "      <td>1.946673e-12</td>\n",
       "      <td>2.557257e-12</td>\n",
       "      <td>2.542725e-05</td>\n",
       "      <td>6.588627e-12</td>\n",
       "      <td>2.334257e-08</td>\n",
       "      <td>1.107869e-09</td>\n",
       "      <td>1.446363e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>2.480479e-09</td>\n",
       "      <td>8.150750e-10</td>\n",
       "      <td>4.013152e-07</td>\n",
       "      <td>3.936375e-12</td>\n",
       "      <td>1.517352e-11</td>\n",
       "      <td>9.978972e-09</td>\n",
       "      <td>3.328797e-12</td>\n",
       "      <td>1.614268e-11</td>\n",
       "      <td>3.113636e-08</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23844 rows × 23844 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ENSMUSG00000000028  ENSMUSG00000000049  \\\n",
       "ENSMUSG00000000028        1.000000e+00        2.737436e-08   \n",
       "ENSMUSG00000000049        2.737436e-08        1.000000e+00   \n",
       "ENSMUSG00000000056        1.547511e-03        6.104462e-08   \n",
       "ENSMUSG00000000058        2.073630e-08        3.473789e-09   \n",
       "ENSMUSG00000000078        4.725714e-09        8.791817e-08   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        3.715675e-10        1.712946e-10   \n",
       "ENSMUSG00000118380        8.465453e-10        3.459822e-10   \n",
       "ENSMUSG00000118382        5.111236e-07        3.414124e-06   \n",
       "ENSMUSG00000118383        2.642267e-05        6.195964e-09   \n",
       "ENSMUSG00000118385        7.450698e-09        1.261908e-13   \n",
       "\n",
       "                    ENSMUSG00000000056  ENSMUSG00000000058  \\\n",
       "ENSMUSG00000000028        1.547511e-03        2.073630e-08   \n",
       "ENSMUSG00000000049        6.104462e-08        3.473789e-09   \n",
       "ENSMUSG00000000056        1.000000e+00        9.395216e-07   \n",
       "ENSMUSG00000000058        9.395216e-07        1.000000e+00   \n",
       "ENSMUSG00000000078        1.383738e-07        1.402169e-04   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        5.748162e-09        1.063537e-09   \n",
       "ENSMUSG00000118380        8.779400e-08        3.631234e-10   \n",
       "ENSMUSG00000118382        7.341468e-06        1.143298e-05   \n",
       "ENSMUSG00000118383        4.213346e-03        1.094910e-07   \n",
       "ENSMUSG00000118385        2.630794e-05        1.946673e-12   \n",
       "\n",
       "                    ENSMUSG00000000078  ENSMUSG00000000085  \\\n",
       "ENSMUSG00000000028        4.725714e-09        2.397450e-03   \n",
       "ENSMUSG00000000049        8.791817e-08        7.308956e-08   \n",
       "ENSMUSG00000000056        1.383738e-07        5.168019e-02   \n",
       "ENSMUSG00000000058        1.402169e-04        4.533858e-07   \n",
       "ENSMUSG00000000078        1.000000e+00        2.399752e-05   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        9.185583e-10        3.497042e-09   \n",
       "ENSMUSG00000118380        1.696247e-09        2.325875e-07   \n",
       "ENSMUSG00000118382        2.139086e-05        7.657257e-05   \n",
       "ENSMUSG00000118383        4.533132e-06        1.784947e-02   \n",
       "ENSMUSG00000118385        2.557257e-12        2.542725e-05   \n",
       "\n",
       "                    ENSMUSG00000000088  ENSMUSG00000000120  \\\n",
       "ENSMUSG00000000028        3.166083e-09        1.218259e-07   \n",
       "ENSMUSG00000000049        5.519291e-07        5.496296e-10   \n",
       "ENSMUSG00000000056        6.775585e-10        1.965049e-04   \n",
       "ENSMUSG00000000058        3.763688e-04        9.523902e-08   \n",
       "ENSMUSG00000000078        6.002748e-03        4.886556e-05   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        2.280045e-09        1.969702e-09   \n",
       "ENSMUSG00000118380        6.027817e-09        2.778879e-08   \n",
       "ENSMUSG00000118382        1.623556e-05        2.521969e-09   \n",
       "ENSMUSG00000118383        1.140550e-09        2.053265e-07   \n",
       "ENSMUSG00000118385        6.588627e-12        2.334257e-08   \n",
       "\n",
       "                    ENSMUSG00000000125  ENSMUSG00000000126  ...  \\\n",
       "ENSMUSG00000000028        5.862595e-07        5.512740e-06  ...   \n",
       "ENSMUSG00000000049        1.428166e-07        4.770658e-08  ...   \n",
       "ENSMUSG00000000056        1.342360e-04        8.450327e-04  ...   \n",
       "ENSMUSG00000000058        1.230038e-08        1.043700e-05  ...   \n",
       "ENSMUSG00000000078        1.993526e-06        3.175543e-03  ...   \n",
       "...                                ...                 ...  ...   \n",
       "ENSMUSG00000118372        3.565670e-11        3.751951e-09  ...   \n",
       "ENSMUSG00000118380        1.148290e-08        4.950899e-12  ...   \n",
       "ENSMUSG00000118382        6.758380e-08        5.919615e-09  ...   \n",
       "ENSMUSG00000118383        1.130612e-05        1.774412e-05  ...   \n",
       "ENSMUSG00000118385        1.107869e-09        1.446363e-06  ...   \n",
       "\n",
       "                    ENSMUSG00000118346  ENSMUSG00000118353  \\\n",
       "ENSMUSG00000000028        1.396855e-05        3.388088e-08   \n",
       "ENSMUSG00000000049        1.827312e-08        1.848953e-12   \n",
       "ENSMUSG00000000056        6.585115e-06        2.099789e-05   \n",
       "ENSMUSG00000000058        5.468722e-03        1.308723e-09   \n",
       "ENSMUSG00000000078        1.299579e-07        8.334004e-13   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        5.720627e-10        8.491071e-09   \n",
       "ENSMUSG00000118380        5.866245e-07        1.493508e-10   \n",
       "ENSMUSG00000118382        9.139048e-06        1.424658e-08   \n",
       "ENSMUSG00000118383        1.389545e-05        6.388611e-07   \n",
       "ENSMUSG00000118385        2.480479e-09        8.150750e-10   \n",
       "\n",
       "                    ENSMUSG00000118369  ENSMUSG00000118370  \\\n",
       "ENSMUSG00000000028        4.887126e-07        5.985277e-12   \n",
       "ENSMUSG00000000049        9.663349e-09        2.443669e-10   \n",
       "ENSMUSG00000000056        8.782336e-04        1.147629e-09   \n",
       "ENSMUSG00000000058        9.309275e-07        1.077963e-08   \n",
       "ENSMUSG00000000078        5.657249e-05        3.574995e-07   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        2.053632e-09        1.828569e-10   \n",
       "ENSMUSG00000118380        7.485294e-11        5.500307e-10   \n",
       "ENSMUSG00000118382        1.356658e-08        2.996882e-08   \n",
       "ENSMUSG00000118383        7.089825e-07        2.036567e-10   \n",
       "ENSMUSG00000118385        4.013152e-07        3.936375e-12   \n",
       "\n",
       "                    ENSMUSG00000118371  ENSMUSG00000118372  \\\n",
       "ENSMUSG00000000028        1.366288e-09        3.715675e-10   \n",
       "ENSMUSG00000000049        3.150939e-10        1.712946e-10   \n",
       "ENSMUSG00000000056        8.615273e-08        5.748162e-09   \n",
       "ENSMUSG00000000058        5.672777e-09        1.063537e-09   \n",
       "ENSMUSG00000000078        9.083470e-09        9.185583e-10   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        4.668191e-12        1.000000e+00   \n",
       "ENSMUSG00000118380        9.493952e-11        1.677540e-10   \n",
       "ENSMUSG00000118382        2.540177e-07        2.121388e-11   \n",
       "ENSMUSG00000118383        9.645142e-08        9.607878e-11   \n",
       "ENSMUSG00000118385        1.517352e-11        9.978972e-09   \n",
       "\n",
       "                    ENSMUSG00000118380  ENSMUSG00000118382  \\\n",
       "ENSMUSG00000000028        8.465453e-10        5.111236e-07   \n",
       "ENSMUSG00000000049        3.459822e-10        3.414124e-06   \n",
       "ENSMUSG00000000056        8.779400e-08        7.341468e-06   \n",
       "ENSMUSG00000000058        3.631234e-10        1.143298e-05   \n",
       "ENSMUSG00000000078        1.696247e-09        2.139086e-05   \n",
       "...                                ...                 ...   \n",
       "ENSMUSG00000118372        1.677540e-10        2.121388e-11   \n",
       "ENSMUSG00000118380        1.000000e+00        3.195543e-06   \n",
       "ENSMUSG00000118382        3.195543e-06        1.000000e+00   \n",
       "ENSMUSG00000118383        8.592797e-08        6.855203e-05   \n",
       "ENSMUSG00000118385        3.328797e-12        1.614268e-11   \n",
       "\n",
       "                    ENSMUSG00000118383  ENSMUSG00000118385  \n",
       "ENSMUSG00000000028        2.642267e-05        7.450698e-09  \n",
       "ENSMUSG00000000049        6.195964e-09        1.261908e-13  \n",
       "ENSMUSG00000000056        4.213346e-03        2.630794e-05  \n",
       "ENSMUSG00000000058        1.094910e-07        1.946673e-12  \n",
       "ENSMUSG00000000078        4.533132e-06        2.557257e-12  \n",
       "...                                ...                 ...  \n",
       "ENSMUSG00000118372        9.607878e-11        9.978972e-09  \n",
       "ENSMUSG00000118380        8.592797e-08        3.328797e-12  \n",
       "ENSMUSG00000118382        6.855203e-05        1.614268e-11  \n",
       "ENSMUSG00000118383        1.000000e+00        3.113636e-08  \n",
       "ENSMUSG00000118385        3.113636e-08        1.000000e+00  \n",
       "\n",
       "[23844 rows x 23844 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyWGCNA_5xFAD.TOM"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
